src/depends/libDatabase/MemoryDB.cpp

```diff
@@ -147,25 +147,51 @@ namespace dev
         m_aux[_h] = make_pair(_v.toBytes(), true);
     }
 
-    void MemoryDB::purge()
+    void MemoryDB::purge(std::vector<h256>& purged, bool haveMutex)
```

Comment:
`haveMutex` name seems confusing because if it's true then you create mutex.
Maybe rename to `calledWithoutMutex` to be clearer.

Also, I suggest to rewrite like this to avoid duplicate code:

```
void MemoryDB::purgeMain(std::vector<h256>& purged)
{
  // code for purge m_main
  // code for purge m_aux
}

void MemoryDB::purge(std::vector<h256>& purged, bool calledWithoutMutex)
{
  if (calledWithoutMutex) {
    unique_lock<shared_timed_mutex> lock(x_this);
    purgeMain(purged);
  }
  else {
    purgeMain(purged);
  }
}
```

Reply:
thx, fixed